Machinery to enable penalty mortar mechanical contact

lindsayad opened the pull request:
Refs #21329

lindsayad committed:
Slight fixes in MortarNodalAux system

lindsayad committed:
First pass at MortarNodalUserObject

Refs #21329

lindsayad committed:
Move user object from framework to contact

lindsayad committed:
Rename to WeightedGapUserObject

lindsayad committed:
Refactor data down into MortarConsumerInterface

lindsayad committed:
Add WeightedGapUserObject

And its derived class LMWeightedGapUserObject

lindsayad committed:
Have user object supply contact force

lindsayad committed:
Add a penalty implementation

lindsayad committed:
Tighten solve tolerances

Seeing exodiffs when running with 16 processes

This makes me nervous

lindsayad committed:
Don't execute user object when we don't have dof

lindsayad committed:
Fix bug in weighted gap computations

In the code we were adding to the dof-to-weighted-gap map from the
same mortar segments multiple times because we were adding to weighted
gaps for nodes different than the current node

This work made me realize that I'd like to make dedicated mortar
user objects that loop over the mortar segments in an elemental way
like our constraints so that we don't do dumb things like linear
searches through element's nodes to figure out whether we should
add to the weighted gap. Additionally even with this first design,
where I *thought* I could get away without doing any communication,
I *do* need communication for the penalty method because when I
actually go to apply the contact force in the displacement mortar
constraints I need to be able to access all the weighted gaps for
all the nodes attached to the current lower secondary element and
those nodes may be off-process

lindsayad committed:
Convert to using element-based mortar user object

lindsayad committed:
Qualify size_t with std::

lindsayad committed:
Simplify mortar contact communication APIs

We don't have to pass the processor ID because we can get that
directly from the communicator

recuero commented on framework/src/loops/ComputeMortarFunctor.C:
```diff
@@ -112,7 +118,10 @@ ComputeMortarFunctor::operator()(const Moose::ComputeType compute_type)
       {
         for (auto * const mc : _mortar_constraints)
+        {
+          mc->setNormals();
```
Mental note: It appears we would have to compute tangents here as well, which will be needed for any kind of frictional modeling/enforcement.

recuero committed:
Add user objects for weighted velocities (mortar friction)

WeightedVelocitiesUserObject and LMWeightedVelocitiesUserObject do the computations for
weighted velocities which are queried by the constraint ComputeFrictionalForceLMMechanicalContact.

One test is converted, which appears to work well in parallel (there was a need to modify
parallel communication for communicating the weighted velocities).

recuero committed:
Adapt and verify contact action with new uo design (frictional test)

recuero committed:
Mechanical mortar contact uses averaged (not interpolated) geometry.

 - It means interpolate_normals is false for all mechanical objects
 - Remove old mortar/nodal mortar objects (i.e. non-weighted quantities objects)
 - Migrate a good number of tests to use frictionless/frictional uo-based design
 - Regolding is necessary for many cases (now we don't interpolate the geometry)

More migration work coming up.

recuero committed:
Finish migrating most of contact module cases

Remain:
 - Frictional 2d cases in mortar_tm
 - Some cases in mortar_aux_kernel. Finding non-deterministic behavior from mortar
   aux kernels. Not sure of the source.

lindsayad committed:
Make WeightedGapVelAux work with non-interpolated normals

lindsayad committed:
Make sure that we don't have the incorrect aux system

lindsayad committed:
Make sure correct variable copy is inserting into solution vector

With the change to make `_var` potentially be the displaced copy,
we need to ensure we insert into the global solution vector using
the correct copy.

This PR also performs some cleanup of code in the aux system and
in the aux variable computation loops

lindsayad committed:
Use _test_lower in WeightedGapAux

lindsayad committed:
Avoid using _coord and _test in MortarNodalAuxKernels

lindsayad committed:
Remove references to AD macros

lindsayad committed:
Prevent indexing _normals OOB in MortarArchardsLawAux

lindsayad committed:
Remove some more ad_indexing_type

recuero committed:
Adjusts tests to run with nodal normals

recuero committed:
Fix non-unity build

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1422288/) on 8e6be0e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23591/site)

This comment will be updated on new commits.

recuero committed:
Migrate combined module tests

recuero committed:
Regold field split (max rel diff 1.0e-3)

recuero committed:
Address parallel failures

lindsayad commented on modules/contact/test/tests/mortar_aux_kernels/tests:
```diff
@@ -90,8 +90,9 @@
-    max_parallel = 4
+    # debug mesh only runs in serial
+    max_parallel = 1
+    max_threads = 1
```
Is there a problem with threads?

recuero replied:
It crashes with threads, but that test has `debug_mesh = true` so it's not intended to work with threads?

lindsayad replied:
Hmm I'm pretty sure you created `debug_mesh` so you'll have to tell me whether it's intended to work with threads haha

lindsayad replied:
If it's not supposed to work with threads then we should add an assertion or something to that effect

recuero replied:
> you'll have to tell me whether it's intended to work with threads haha

That capability uses  `write_equation_systems` in libMesh. It may be able to run in parallel but I have never used it to that effect. I think I was mostly thinking of the mortar statistics --which is handled via the same boolean, but that shouldn't make the test fail.

recuero replied:
I am reverting that change.

recuero committed:
Make finite_rr_full converge better (multiple-projections-lowerd remains).

recuero committed:
Fix test with no projection.

While migrating cases, I turn this lets-not-solve-edge correctly to
a lets-solve-edge correctly inadvertently. This problem does not converge
well when treating the contact patch edge properly. The influence of edge
dropping is known to mess with the condition number. In this case, while
the condition number with and without edge dropping treatment is similar,
convergence changes radically. So we will need to provide an alleviating
technique in addition to condensation, which we haven't used extensively
yet.

lindsayad committed:
Only compute if _var.isNodalDefined() in MortarNodalAuxKernel

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1426394/) on cf6dc1e wanted to post the following:


The following coverage requirement(s) failed:
- `contact` coverage rate 81.93% is less than the required 87.0%

lindsayad committed:
We now only support nodal variables in weighted objects

lindsayad committed:
Add framework test for mortar user objects

lindsayad committed:
Remove unused constraint objects

lindsayad committed:
Remove send-data-back code in communicateVelocities

lindsayad committed:
Ensure coverage of back-communication of weighted gaps

lindsayad committed:
MortarUserObjects don't support writable coupled variables

So don't add unused code for it

lindsayad committed:
Add displaced mesh framework test

recuero commented on framework/src/systems/AuxiliarySystem.C:
```diff
@@ -52,13 +52,7 @@ AuxiliarySystem::AuxiliarySystem(FEProblemBase & subproblem, const std::string &
     _elemental_array_aux_storage(_app.getExecuteOnEnum())
 {
   _nodal_vars.resize(libMesh::n_threads());
-  _nodal_std_vars.resize(libMesh::n_threads());
-  _nodal_vec_vars.resize(libMesh::n_threads());
-  _nodal_array_vars.resize(libMesh::n_threads());
   _elem_vars.resize(libMesh::n_threads());
-  _elem_std_vars.resize(libMesh::n_threads());
-  _elem_vec_vars.resize(libMesh::n_threads());
-  _elem_array_vars.resize(libMesh::n_threads());
```
Nice consolidation.

recuero replied:
But honestly, I don't fully understand why they were needed before. They seemed to be holding the same variables.

lindsayad replied:
They were used in the unnecessary calls to things like `prepareAux` and also for things like `var->insert` ... but we don't need the former and the latter is better accomplished by calling `kernel->variable().insert`. Why is it better? Because the variable may be in the displaced aux system and not the undisplaced one

recuero replied:
Okay, circling back to your other comment on `aux_sys` acting working only for undisplaced meshes.

recuero commented on modules/combined/test/tests/gap_heat_transfer_mortar/finite-2d-rz/finite.i:
```diff
@@ -122,6 +122,19 @@ name = 'finite'
+    primary_subdomain = frictionless_primary_subdomain
+    secondary_subdomain = frictionless_secondary_subdomain
+    lm_variable = frictionless_normal_lm
+    disp_x = disp_x
+    disp_y = disp_y
```
@gambka mentioned that the latest convention in MOOSE and apps is to use `displacements` instead of the individual components --which I think we've been inheriting from original mortar objects. We can make that change in follow-up PR when we find some time.

recuero commented on modules/combined/test/tests/gap_heat_transfer_mortar/small-2d-rz/small.i:
```diff
@@ -147,6 +161,7 @@ name = 'small'
     component = x
     use_displaced_mesh = true
```
Another follow-up change we could make if you agree is to default these "apply-contact-force" objects to use the displaced mesh. I think that'd be the right setting.

recuero commented on modules/contact/include/userobjects/LMWeightedGapUserObject.h:
```diff
@@ -0,0 +1,36 @@
+class MooseVariableFE;
+
+/**
+ * Base class for creating new nodally-based mortar user objects
```
Not the base class

recuero commented on modules/contact/include/userobjects/LMWeightedGapUserObject.h:
```diff
@@ -0,0 +1,36 @@
+
+  LMWeightedGapUserObject(const InputParameters & parameters);
+
+  virtual const ADVariableValue & contactPressure() const override;
```
The power of AD for penalty!!!

recuero commented on modules/contact/include/utils/MortarContactUtils.h:
```diff
@@ -92,19 +94,21 @@ communicateVelocities(std::unordered_map<const DofObject *, std::array<T, 2>> &
  * @param nodal Whether the element has Lagrange interpolation
  * @param normalize_c Whether to normalize with size the c coefficient in contact constraint
  * @param communicator Process communicator
+ * @param send_data_back After aggregating data on the owning process, whether to send the aggregate
+ * back to senders. This can be necessary for things like penalty contact in which the constraint is
+ * not enforced by the owner but in a weighted way by the displacement constraints
```
👍

recuero commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -96,184 +101,77 @@ ComputeFrictionalForceLMMechanicalContact::ComputeFrictionalForceLMMechanicalCon
-    _weighted_gap_ptr = &weighted_gap_pr.first;
-    _normalization_ptr = &weighted_gap_pr.second;
-    _tangential_vel_ptr[0] = &(pr.second[0]);
+    // Substitute 'at' for libmesh find counterpart per MOOSE people directive.
```
I need to address that.

lindsayad replied:
I went ahead and made this change

recuero replied:
Thanks!!

recuero commented on modules/contact/src/userobjects/WeightedGapUserObject.C:
```diff
@@ -0,0 +1,155 @@
+}
+
+WeightedGapUserObject::WeightedGapUserObject(const InputParameters & parameters)
+  : MortarUserObject(parameters),
```
General comment. We have now another layer of mortar objects which draw on base class functionality. Stuff like `correct_edge_dropping` and `minimum_projection_angle` (or was it maximum?) will need to be synchronize among these uos and constraints. Maybe a good opportunity to add more functionality to the action and not encourage the user to set things up by themselves.

recuero replied:
Okay, gather all these things into a small issue. At least, I won't forget: https://github.com/idaholab/moose/issues/23889

recuero commented on test/src/userobjects/TestWeightedGapUserObject.C:
```diff
@@ -0,0 +1,142 @@
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
```
Do we need these test objects for coverage?

lindsayad replied:
yea otherwise we don't have any mortar user object usage for framework coverage testing

lindsayad commented on framework/include/interfaces/MortarConsumerInterface.h:
```diff
@@ -70,9 +74,19 @@ class MortarConsumerInterface
+   * Get rid of AD derivative entries by dof index
    */
-  void setNormals();
+  static void trimDerivative(dof_id_type remove_derivative_index, ADReal & dual_number);
```
Any inheritor from this interface can call this method. And I don't that's a bad thing. Even a non-AD mortar constraint could choose to do some calculations with AD if they wanted to

lindsayad commented on framework/include/materials/MaterialBase.h:
```diff
@@ -465,3 +474,58 @@ MaterialBase::declareADPropertyByName(const std::string & prop_name_in)
+    needed_mat_props.insert(mp_deps.begin(), mp_deps.end());
+  }
+
+  // A predicate of calling this function is that these materials come in already sorted by
```
I could add a long immediately invoked lambda that calls the dependency resolver and ensures that what comes back matches the incoming ordering

lindsayad commented on framework/include/variables/MooseVariableFE.h:
```diff
@@ -237,7 +237,7 @@ class MooseVariableFE : public MooseVariableField<OutputType>
   const FieldVariablePhiCurl & curlPhiFaceNeighbor() const;
 
-  const FieldVariablePhiValue & phiLower() const { return _lower_data->phi(); }
+  virtual const FieldVariablePhiValue & phiLower() const override { return _lower_data->phi(); }
```
The mortar system works with finite volume variables. And we call `_var.phiLower()` in `MortarNodalAuxKernel` (now) where `_var` is a `MooseVariableField` type

lindsayad commented on framework/src/auxkernels/MortarNodalAuxKernel.C:
```diff
@@ -79,6 +82,9 @@ template <typename ComputeValueType>
 void
 MortarNodalAuxKernelTempl<ComputeValueType>::compute()
 {
+  if (!_var.isNodalDefined())
```
Yes. This could theoretically happen even for a serial run if you have a second order mesh and your aux variable is first order

lindsayad commented on framework/src/base/MooseObject.C:
```diff
@@ -44,7 +44,6 @@ MooseObject::validParams()
   params.addPrivateParam<SubProblem *>("_subproblem", nullptr);
   params.addPrivateParam<SystemBase *>("_sys", nullptr);
   params.addPrivateParam<SystemBase *>("_nl_sys", nullptr);
-  params.addPrivateParam<AuxiliarySystem *>("_aux_sys", nullptr);
```
In the past `FEProblemBase` was assigning the undisplaced auxiliary system as `_aux_sys` even when an auxiliary kernel was acting on the displaced mesh. This was wrong and lead to problems earlier in the PR. So I've removed this parameter to avoid the possibility of people doing this stupid thing again. The assignment of `_sys` has always been correct

recuero replied:
Great. I'd have definitely fallen for that.

lindsayad commented on framework/src/loops/ComputeElemAuxBcsThread.C:
```diff
@@ -66,10 +63,6 @@ ComputeElemAuxBcsThread<AuxKernelType>::operator()(const ConstBndElemRange & ran
 
     if (elem->processor_id() == _problem.processor_id())
     {
-      // prepare variables
```
It's already done in either the call to `_problem.prepare` or `_problem.reinitElemFace`

lindsayad commented on framework/src/loops/MortarUserObjectThread.C:
```diff
@@ -0,0 +1,96 @@
+    for (auto * const mc : _mortar_user_objects)
+    {
+      mc->setNormals();
+      mc->execute();
```
The issue is that `doMortarThings()` is always a virtual function, e.g. `computeResidual`, `computeJacobian`, `execute` etc. It's very possible for someone overriding those virtuals to forget the call to `setNormals`. So what I've done in this PR is to make `setNormals` private and limited its use by declaring the four classes that need to call it to make calculations right as friends of `MortarConsumerInterface`. This ensures that we keep `setNormals` in the framework so we have tight control

lindsayad commented on framework/src/relationshipmanagers/AugmentSparsityOnInterface.C:
```diff
@@ -222,22 +204,25 @@ AugmentSparsityOnInterface::ghostLowerDSecondaryElemPointNeighbors(const process
 void
 AugmentSparsityOnInterface::ghostHigherDNeighbors(const processor_id_type p,
```
The writing of the member variables is not thread safe. And `operator()` is called in a threaded fashion when we are computing sparsity. I discovered this thread unsafety when debugging some other thread issue during this PR

lindsayad committed:
Address Antonio review

- Fix doxygen for my weighted gap user objects
- Use `libmesh_map_find` instead of `at`

lindsayad commented on framework/include/materials/MaterialBase.h:
```diff
@@ -465,3 +474,58 @@ MaterialBase::declareADPropertyByName(const std::string & prop_name_in)
+
+    if (!allow_stateful && mat->hasStatefulProperties())
+      ::mooseError(
+          "Someone called buildRequiredMaterials with allow_stateful = false but a material "
```
Yea we could definitely explore it if/when the need arises

recuero commented:
I am approving @lindsayad's code within this PR.

dschwen commented on modules/contact/src/userobjects/WeightedGapUserObject.C:
```diff
@@ -0,0 +1,155 @@
+{
+  // If the constraint is performed by the owner, then we don't need any data sent back; the owner
+  // will take care of it. But if the constraint is not performed by the owner and we might have to
+  // do some of the constraining ourselves, then we need data sent back to us
```
This is not clear to me. Why would anyone but the owner perform the constraining?

lindsayad replied:
The penalty constraint is a distributed constraint. You may compute a weighted gap for Node 0 owned by processor 0, and that node is attached to an element on processor 0 and another element on processor 1. The penalty constraint driving the weighted gap on node 0 to zero will be applied through the contact force applied to both the element on processor 0 and the element on processor 1. Whereas for a Lagrange multiplier implementation the constraint is enforced in the Lagrange multiplier constraint object itself and only on a local basis

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1451062/) on 0204388 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23591/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23591/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4792096a38985305588ca481a6e55174cee9f084`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1451531/) on ed3c10b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23591/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23591/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format aa09c949d20d5f5a4a446f5e5b9e0cf5296b4a0d`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 06b0b74d48cb9ab8152c11db348caa530810a7c9
+Subproject commit 0b70d9f5766eb7b0bd56c2716ff6c7d6c4c87c7c
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1451617/) on 73e4cc1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23591/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23591/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format aa09c949d20d5f5a4a446f5e5b9e0cf5296b4a0d`

recuero committed:
Fix merge and clang format

dschwen commented on framework/doc/content/syntax/UserObjects/index.md:
```diff
@@ -36,6 +36,8 @@ and use the interface without the use of dynamic casts.
 - +DomainUserObject+: this object is capable of executing all the operations of
   a +ElementUserObject+, +InternalSideUserObject+, +SideUserObject+ and +InterfaceUserObject+.
+- +MortarUserObject+: "execute" is called for each mortar segment element corresponding to the
+  secondary-primary mortar interface pairing on each execution flag
```
`MortarUserObject`s should be added to the section below as well, but we can fix this later

dschwen commented on framework/include/auxkernels/MortarNodalAuxKernel.h:
```diff
@@ -61,6 +61,25 @@ class MortarNodalAuxKernelTempl : public AuxKernelTempl<ComputeValueType>,
+  const MooseArray<Real> & _coord_msm;
+
+private:
+  // These are invalid for use because they are only meant for working with the higher-dimensional
+  // elements
```
so you are changing the access qualifier here to prevent use by derived classes?

dschwen commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -742,19 +808,22 @@ ContactAction::addMortarContact()
                                     action_name + "_normal_constraint_",
                                     "NormalMortarMechanicalContact",
-                                    false);
+                                    false,
+                                    true);
```
I would love this, but the lambda is defined not far away, so its really just a weak suggestion
```suggestion
                                    /* is_additional_frictional_constraint = */ false,
                                    /* is_normal_constraint = */ true);
```

dschwen replied:
in fact, let's not make any change to PR right now

recuero commented on modules/contact/src/constraints/ComputeDynamicWeightedGapLMMechanicalContact.C:
```diff
@@ -42,13 +55,41 @@ ComputeDynamicWeightedGapLMMechanicalContact::validParams()
     const InputParameters & parameters)
-  : ComputeWeightedGapLMMechanicalContact(parameters),
+  : ADMortarConstraint(assignVarsInParamsDyn(parameters)),
+    _secondary_disp_x(adCoupledValue("disp_x")),
+    _primary_disp_x(adCoupledNeighborValue("disp_x")),
+    _secondary_disp_y(adCoupledValue("disp_y")),
+    _primary_disp_y(adCoupledNeighborValue("disp_y")),
+    _has_disp_z(isCoupled("disp_z")),
+    _secondary_disp_z(_has_disp_z ? &adCoupledValue("disp_z") : nullptr),
+    _primary_disp_z(_has_disp_z ? &adCoupledNeighborValue("disp_z") : nullptr),
```
Already on my plate

https://github.com/idaholab/moose/issues/23889

dschwen commented on modules/contact/src/userobjects/LMWeightedGapUserObject.C:
```diff
@@ -0,0 +1,48 @@
+LMWeightedGapUserObject::LMWeightedGapUserObject(const InputParameters & parameters)
+  : WeightedGapUserObject(parameters), _lm_var(getVar("lm_variable", 0))
+{
+  if (!_lm_var)
+    paramError("lm_variable",
+               "The Lagrange multiplier variable must be an actual variable and not a constant.");
```
`_lm_var` being `nullptr` could also mean that the variable does not exist! (e.g. the user made a typo). We have `isCoupledConstant` now (added in #22598) to disambiguate those cases ... if you want to improve the error in the future.
